### DISCUSSION TITLE
TipMass setting

### DISCUSSION THREAD

**Discussion by TheKite4** [OpenFAST/openfast#1499] [Q&A]
Hi,

I tried to simulate a scaled-down WindFloat semi-submersible by OpenFAST downloaded from here https://github.com/openfast/openfast/releases/tag/v3.4.1.

To simplify the simulation, here is my modules setting: 
![image](https://user-images.githubusercontent.com/93707978/226196238-eecaceb0-2b6b-411c-bd1e-7e94bd39b3d0.png)

But it shows an error:
FAST_InitializeAll:ED_Init:ED_ReadInput:ReadPrimaryFile:Invalid numerical input for file
".\wow_ElastoDyn.dat" occurred while trying to read TipMass(1).

I know that it means the Tip brake mass, but I don;t think the scaled down wind turbine with 0.645m radius will have this, so I set it to all three of TipMass to be 0. I have no clue what goes wrong with this. Here is my 'MASS AND INERTIA' settings in ElastoDyn (I set the turbine rotation mass and inertia extreme high to assume the wind turbine rotation dof is fixed):

          0   TipMass(1)  - Tip-brake mass, blade 1 (kg)
          0   TipMass(2)  - Tip-brake mass, blade 2 (kg)
          0   TipMass(3)  - Tip-brake mass, blade 3 (kg) [unused for 2 blades]
        0.6   HubMass     - Hub mass (kg)
  0.0005625   HubIner     - Hub inertia about rotor axis [3 blades] or teeter axis [2 blades] (kg m^2)
          10000000   GenIner     - Generator inertia about HSS (kg m^2)
        5.5   NacMass     - Nacelle mass (kg)
          10000000   NacYIner    - Nacelle inertia about yaw axis (kg m^2)
          10000000   YawBrMass   - Yaw bearing mass (kg)
       22.3   PtfmMass    - Platform mass (kg)
	   5   PtfmRIner   - Platform inertia for roll tilt rotation about the platform CM (kg m^2)
          8.6 PtfmPIner   - Platform inertia for pitch tilt rotation about the platform CM (kg m^2)
          3.6  PtfmYIner   - Platform inertia for yaw rotation about the platform CM (kg m^2)

**Comment by jjonkman** [comment]
Dear @TheKite4,

I would guess when you edited the ElastoDyn input file, that the format of the file was accidently changed (e.g., with a line removed or added).  As with any input file processing error, I would enable the `Echo` option to debug.

FYI: A couple more comments:

- It is much more common to represent a model-scale wind turbine in full-scale and scale up model-scale experimental data (such as wind tunnel data or wave tank data) to full scale for comparison..  Some of the built-in tolerances within OpenFAST may need to be changed if you implement an OpenFAST model at model scale.
- I would not suggest setting mass or inertias to be very high because this may result in ill-conditioning of the ElastoDyn mass matrix.  You can always disable structural degrees of freedom in ElastoDyn that are unwanted.

Best regards,

**Reply by TheKite4** [reply]
Hi jjonkman:

I scaled up my model to 5MW and replace the HydroDyn in a 5MW OC4 semi-submersible example, to check my full-scale performance. However, I am not sure whether the outputs are correct. 
![0675bc27bac8cac39ca1c34a3e0011f](https://user-images.githubusercontent.com/93707978/227285397-3fc59b9d-e71f-487d-a678-a4bd0fa311b2.png)
As my model is a TLP, I think it makes sense that the pitch and heave reduces a lot compared to the original OC4 semi-submersible performance. Could you help me confirm, or is there any other simulation setting I might need to check?

**Reply by jjonkman** [reply]
Dear @TheKite4,

I would say that at least your heave and pitch do not look correct.  For heave, the HydroDyn modeling guidance (https://openfast.readthedocs.io/en/main/source/user/hydrodyn/modeling_considerations.html#undisplaced-position-for-floating-systems) recommends that you set up your model to have zero heave displacement in the absence of wind or wave excitation. I'm not sure what conditions you are simulating, but the heave displacement looks large.  For pitch, I would expect that a TLP would have negligible pitch rotation (likely less than a degree under all conditions).

Best regards,

**Reply by TheKite4** [reply]
Dear @jjonkman:

These days I am debugging my small-scale model, and I am now stuck in the error found in MoorDyn, it appears some extreme calculation errors, which are:
![image](https://user-images.githubusercontent.com/93707978/227805935-c97abd66-b870-479d-a871-ded8b47c758b.png)
I have varied my mooring line settings and checked the posts of this problem before, but I didn't find any solution. I guess this is the "Some of the built-in tolerances within OpenFAST may need to be changed if you implement an OpenFAST model at model scale." you suggested before? 

I am still checking other settings, could you tell me if you have any idea, please? And here is my small-scale MoorDyn file:
![image](https://user-images.githubusercontent.com/93707978/227806145-5cb4ef90-de84-40e3-bbe1-45eca631d51d.png)

Cheers.

**Reply by mattEhall** [reply]
Hi. First off, make sure you do the usual adjustments for model stability problems (i.e. see if you can solve the instability by reducing dtM enough). But if that doesn't solve things, I'm not sure what to suggest aside from either scaling up or artificially reducing the system stiffness by lowering the EA value. For reasons I don't fully understand, sometimes simulations at model scale have unique difficulties, as Jason suggested.
Matt

**Comment by TheKite4** [comment]
Hi, I have tried to use MoorDyn but somehow it doesn't work though I have checked the input files. I decided to directly use AddF0 and stiffness matrix in HydroDyn. I fix the 3 rotation motions in Elasto to be False because I found it triggered a "too large platform angle displacement". Now it runs. but when I checked the output, it's all NaN. Could anyone help me with this? Thanks!
![image](https://user-images.githubusercontent.com/93707978/228977322-7771658b-92d0-4f1e-9a75-a4b94333342d.png)
![image](https://user-images.githubusercontent.com/93707978/228977366-45edc44a-952d-4acd-b20b-93ec1b592f45.png)

**Reply by jjonkman** [reply]
Dear @TheKite4,

You haven't provided enough information to know why, but it looks like your OpenFAST model is unstable.

Best regards,

**Reply by TheKite4** [reply]
Dear @jjonkman,

Here is a layout of the model set, if you need more information please let me know.
1. As the wind turbine I have doesn't have a control mechanism, I didn't include a ServoDyn. For my TLP design, I have tried to use MoorDyn but it doesn't work, so I chose to directly input AddF0_y and a stiffness matrix in HydroDyn. It turns out to be at least "terminated normally", but there was a warning stated below, and the .out file is almost all NaN as shown before.

SetCoordSy:Small angle assumption violated in SUBROUTINE SmllRotTrans() due to a large platform displacement (ElastoDyn
SetCoordSy). The solution may be inaccurate. Simulation continuing, but future warnings from SmllRotTrans() will be suppressed.

Then I followed the advice from some other posts about similar issues. I turned off the rotation dof in ElastoDyn. Though there is no that warning this time (shown in the screensnap before), I got a same .out file with lots of NaN. I have calculated the metacentric height to be larger than 0. Here is my platform force balance-related settings.

From ED_sum, the mass of the platform:
![image](https://user-images.githubusercontent.com/93707978/228985571-6aa6a4d9-697a-4ba0-b4e3-7480795b16a5.png)

From HD_sum, the buoyancy setting:
![image](https://user-images.githubusercontent.com/93707978/228985815-d1d90e0f-5b64-46c0-b1d2-31a40b8f3dfc.png)

For the added force and stiffness matrix from the mooring line/tension legs:
![image](https://user-images.githubusercontent.com/93707978/228986270-83cff618-87cc-4e1a-898b-676fcaa5749a.png)

From the warning, it looks like the body is not balanced in rotation degree of freedom, so it instantly triggered the warning in the start of the simulation I think. But I don't know why it still fails after I turned off the rotation dofs of the platform.

If you need more information please let me know, I will respond ASAP.

**Reply by jjonkman** [reply]
Dear @TheKite4,

Just a few comments:
- Why did you specify `AddF0` in the -Y direction and not the -Z direction?
- Your external buoyancy doesn't appear to balance with the system weight and ballast, so, the system will sink.
- Because you appear to simulating at model scale, the natural frequencies will be very high, requiring a minuscule time step.  Have you used a small time step following the modeling guidance based on the highest natural frequency in the model?

Best regards,

**Reply by TheKite4** [reply]
Dear @jjonkman,

The time step I set is very small (0.0001), which I think should be small enough. I have changed AddF0 to -Z direction but it still fails. For the second point, I just want to confirm: does the platform mass set in ElastoDyn include ballast too? If not, I think the platform inertias of the 3 rotation degrees also don't include ballast?

Thank you so much for the help!

**Reply by jjonkman** [reply]
Dear @TheKite4,

Per my second point above, the external buoyancy doesn't seem large enough to support the weight and ballast; now adding the vertical mooring pretension will only exaggerate the issue.  You should double check the vertical equilibrium of your system.

The platform mass, center of mass, and inertias specified in ElastoDyn should only include ballast if the ballast is not already considered in HydroDyn (so as not to double count the ballast).

If your system is still unstable and you feel the time step is sufficient, I would suggest disabling DOFs (down to only a single DOF if needed, such as only platform heave), assessing the results, and adding complexity in steps to debug the issue.

Best regards,

**Comment by TheKite4** [comment]
Hi @jjonkman:

I think the problem is solved now, however, I am not sure whether the output is correct, as I observed some black areas in the figures, especially in pitch and roll motions. Additionally, the rotation is surprisingly small, is the unit rad? I wander if it's because I set the mooring too stiff. 
<img width="495" alt="image" src="https://user-images.githubusercontent.com/93707978/231538140-4df87ba5-430a-482d-8cb6-2c0bea11295c.png">
The water depth is 1.2m with wave height 0.4 m wave period 4s.

**Reply by jjonkman** [reply]
Dear @TheKite4,

The responses look OK to me without knowing anything about the system you are simulating.   Whether "black areas" are correct likely depends on the system properties.  Do you know what frequency content you are expecting to get?  You could calculate the PSD of these time series to see what frequency content is in the OpenFAST solution for comparison.

If you you are plotting the platform motion outputs from the ElastoDyn module, these motions are in meters for translations and degrees for rotations.

Best regards,

**Comment by TheKite4** [comment]
Hi,

I am solving these problems these days, and I have some other questions when reading the Modeling Consideration in HydroDyn documentation:

1. When using WAMIT to output [.1] [.3] and [.hst] files, I notice that it shouldn't include body and mooring restoring to get the [.hst] file, which I assume means I shouldn't include the mass matrix and stiffness matrix in the [.frc] WAMIT input file to get all these three files? Or should I add those two matrices in a separate WAMIT simulation to get [.1] and [.3]?
2. When I set still water surface in HydroDyn and 0 for InflowWind in [.fst] to see the free decay of the platform, it will automatically move with a surge of about -0.5m, but I am not sure why as it's not in the vertical direction.
![image](https://user-images.githubusercontent.com/93707978/233196706-047c3c9e-dde9-47e5-9dd4-b638de875c6a.png)
3. Is there any difference if I don't use MoorDyn but directly add a mooring stiffness matrix in HydroDyn?

Thank you so much!

**Reply by jjonkman** [reply]
Dear @TheKite4,

Here are my responses:
1. The body mass and mooring matrix should have no effect on the first-order wave-radiation (_.1_) solution or the first-order wave-excitation (_.3_) solution.  These will, however, effect the hydrostatic (_.hst_) solution.  So, the HydroDyn model guidance (https://openfast.readthedocs.io/en/main/source/user/hydrodyn/modeling_considerations.html#floating-systems-modeled-with-potential-flow) recommends zeroing out these contributions when solving the first-order problem. 
2. I'm not sure what OpenFAST model you are using (or if you generated your own), but often a floating offshore wind turbine will have a nonzero mean pitch and surge due to the overhanging weight of the rotor-nacelle assembly.
3. An added stiffness matrix (and preload vector) should be able to capture the first-order effect of the mooring system, but will miss mooring system nonlinearities and mooring dynamics.

Best regards,

**Reply by TheKite4** [reply]
Hi @jjonkman:

Thanks for these tips. For the second point, yes, I customized a model in OpenFAST.  The model has a similar shape to the WindFloat semi-submersible, the wind turbine is not in the mid of the triangle platform but on a column of the three. It doesn't have heave plates, and the mooring system is tension legs. Some important parameters to refer the design are:

Wind turbine rotor diameter: 1.2m
Tower height: 0.6m
Platform diameter: 1.2m
Column diameter: 0.25m
Column height: 0.5m
Draft: 0.25m
Water depth: 1.2m

From your second comment, is it possible that my static surge (about -0.55m in the figure) results from the position of the wind turbine which is not in the middle?

**Reply by jjonkman** [reply]
Dear @TheKite4,

It sounds like your are modeling floating wind turbine at model scale.  At the scale you are simulating, a surge of -0.55 m sounds large.  Given that the turbine is installed on an offset column of the semi, is the semi asymmetrical such that installing the turbine on an offset column is not expect to result in much floater pitch and roll rotation?  And given that you are modeling tendons, presumably you are trying to minimize floater pitch and roll rotations; are these angles remaining small?  Is floater heave what you expect?

Best regards,

**Reply by TheKite4** [reply]
Hi @jjonkman,

here is the dynamic response, I set the wave as a regular and unidirectional wave with a height of 0.4m and period 4s.:
![image](https://user-images.githubusercontent.com/93707978/233654960-43676125-5903-4237-8a78-f230dddb9395.png)
I indeed want to minimize the rotations, but I didn't expect them to be such small, which looks like the platform only moves back and forward (surge) without any other rotations. The heave looks fine to me as the system is fixed by mooring tendons. I am still not sure why there is an offset at the surge, and the surge motion is huge.

**Reply by jjonkman** [reply]
Dear @TheKite4,

Indeed, all of the other floater motions look small.  I'm not sure why your surge is so high.

Please note that a wave height of 0.4 m sounds excessively high for your model scale floating wind turbine.

Best regards,

**Reply by TheKite4** [reply]
Hi @jjonkman,

I am trying to use only strip theory to run HydroDyn, however, the surge motion looks not converged. My model is the same one described previously, here are the surge motion and my strip settings:
![image](https://user-images.githubusercontent.com/93707978/233870648-4e130e4c-7b36-42f2-b25c-46602e9b7d05.png)
![image](https://user-images.githubusercontent.com/93707978/233870728-8f387b65-57e3-4443-bdc1-7d9065045a71.png)
![image](https://user-images.githubusercontent.com/93707978/233870751-20328422-44bf-4187-b86e-0a5e8d8cd09d.png)

**Reply by jjonkman** [reply]
Dear @TheKite4,

You mention that your model is a strip-theory only model, but I see that you've specified values for `AddCLin`.  `AddCLin` is only used when `PotMod` = 1.   But presumably you've set `PotMod` = 0 for a strip-theory only solution.  Perhaps that is the problem?

Best regards,

**Reply by TheKite4** [reply]
Hi @jjonkman,

Thanks, it has been solved. Could you please explain the difference between the hydrodynamic coefficients defined in AXIAL COEFFICIENTS and SIMPLE HYDRODYNAMIC COEFFICIENTS? I have read the documentation but am still not sure about it. Additionally, is there any guidance for the coefficient values? In the HydroDyn modelling consideration it's said to set all of them 1, but then my simulation failed.

Cheers.

**Reply by jjonkman** [reply]
Hi @TheKite4,

The axial coefficients are used for axial effects at member ends (at the joints).  The other coefficients are used for transverse effects distributed along each member.  I would say that 1.0 is a good guess for most coefficients in the absence of more specific information, but certainly, the coefficients depend on many things (like Re and KC numbers) and it is better to calibrate against experimental data or CFD if you have such data available.  The axial drag coefficient in (`AxCd`) tends be much higher than 1.0 for heave plates.

Best regards,

**Reply by TheKite4** [reply]
Hi @jjonkman,

Thanks. Just to confirm, is the "axial" direction the direction where the wave goes along the member section? If this is true, then I think the "transverse" direction is the opposite direction. In my case where a unidirectional wave with only x-axis direction is adopted, I only apply axial coefficients. Am I right?

Cheers.

**Reply by jjonkman** [reply]
Hi @TheKite4,

By "axial", I meant along the member, regardless of the orientation of the member relative to the waves.  By "transverse", I meant normal to the member.

Best regards,

**Reply by TheKite4** [reply]
Hi @jjonkman,

I think that problem has been solved. My other question is that my rotor blade seems not to interact with wind speed, it's fixed at the values given in the ElastoDyn. I am not sure why, is it because it requires ServoDyn to make it variable?

Cheers

**Reply by jjonkman** [reply]
Sounds like you have the generator DOF disabled in ElastoDyn.  When you enable the generator DOF (`GenDOF` = True), you'll likely need a controller enabled in ServoDyn to regulate the speed.

Best regards,
